web-server/src/content/Service/SystemLogs.tsx

```diff
@@ -16,82 +16,171 @@ import { parseLogLine } from '@/utils/logFormatter';
 
 import { MotionBox } from '../../components/MotionComponents';
 
-export const SystemLogs = ({ serviceName }: { serviceName?: ServiceNames }) => {
+const SystemLogs = memo(({ serviceName }: { serviceName?: ServiceNames }) => {
   const { services, loading, logs } = useSystemLogs({ serviceName });
-  const showScrollDownButton = useBoolState(false);
   const containerRef = useRef<HTMLDivElement>(null);
+  const showScrollDownButton = useBoolState(false);
+  const [searchQuery, setSearchQuery] = useState('');
+  const [debouncedSearchQuery, setDebouncedSearchQuery] = useState('');
+  const [currentMatch, setCurrentMatch] = useState(0);
+  const [totalMatches, setTotalMatches] = useState(0);
+  const [highlightedElements, setHighlightedElements] = useState<HTMLElement[]>([]);
+  const currentHighlightRef = useRef<HTMLElement | null>(null);
+  const isInitialLoad = useRef(true);
+  const isSearchingRef = useRef(false);
+  const currentMatchRef = useRef(0);
+  const totalMatchesRef = useRef(0);
+  const highlightedElementsRef = useRef<HTMLElement[]>([]);
+  const isNavigatingRef = useRef(false);
 
-  const scrollDown = () => {
-    containerRef.current.scrollIntoView({ behavior: 'smooth' });
-  };
-
+  // Debounce search query updates
   useEffect(() => {
-    const observer = new IntersectionObserver(
-      (entries) => {
-        entries.forEach((entry) => {
-          if (entry.isIntersecting) {
-            showScrollDownButton.false();
-          } else {
-            showScrollDownButton.true();
-          }
-        });
-      },
-      {
-        threshold: 0
-      }
-    );
-
-    const containerElement = containerRef.current;
+    const timer = setTimeout(() => {
+      setDebouncedSearchQuery(searchQuery);
+    }, 300);
+    return () => clearTimeout(timer);
+  }, [searchQuery]);
 
+  const scrollToBottom = useCallback(() => {
     if (containerRef.current) {
-      observer.observe(containerElement);
+      containerRef.current.scrollTo({
+        top: containerRef.current.scrollHeight,
+        behavior: 'smooth'
+      });
+    }
+  }, []);
+
+  const updateHighlight = useCallback((element: HTMLElement | null) => {
+    if (currentHighlightRef.current) {
+      currentHighlightRef.current.style.backgroundColor = 'yellow';
+      currentHighlightRef.current.style.color = 'black';
+    }
+
+    if (element) {
+      element.style.backgroundColor = 'rgb(255, 161, 10)';
+      element.style.color = 'white';
+      currentHighlightRef.current = element;
+    } else {
+      currentHighlightRef.current = null;
     }
```

Comment:
Well... that's not a great idea my dude. Use CSS classes, or even MUI's css-in-js based on some supplied prop. Don't mutate the DOM like this unless there's no other way (sometimes that is the case, but it's not right now).

If you're not sure why I'm saying this, feel free to ask. I urge anyone working with us to not just absorb feedback if they aren't convinced.